server/lib/service/reporting.dart

```diff
@@ -0,0 +1,125 @@
+import 'dart:async';
+
+import 'package:csv/csv.dart';
+import 'package:fixnum/fixnum.dart';
+import 'package:googleapis/doubleclickbidmanager/v1_1.dart';
+import 'package:http/http.dart';
+
+/// A class to interact with the DV360 reporting API.
+class ReportingClient {
+  /// A map from the metric name to the name of the column name in the report.
+  ///
+  /// See: https://developers.google.com/bid-manager/v1.1/filters-metrics
+  static const metricToReportColumnName = {
+    'METRIC_REVENUE_ECPM_ADVERTISER': 'Revenue eCPM (Adv Currency)',
+  };
+
+  /// The Reporting API.
+  final DoubleclickbidmanagerApi _api;
+
+  /// Creates an instance of [ReportingClient].
+  ReportingClient(Client client, String baseUrl)
+      : _api = DoubleclickbidmanagerApi(client, rootUrl: baseUrl);
+
+  /// Gets the CPM of the line item with [advertiserId] and [lineItemId].
+  ///
+  /// Throws an [ApiRequestError] if there is an error with the reporting API.
+  /// Throws an [ArgumentError] if any data is missing from the report or the
+  /// report is not found. Returns the CPM value as a double.
+  Future<double> getLineItemCpm(Int64 advertiserId, Int64 lineItemId) async {
+    return double.parse(await _getLineItemMetric(
+        advertiserId, lineItemId, 'METRIC_REVENUE_ECPM_ADVERTISER'));
+  }
+
+  /// Gets the previous day metric value of the line item with [advertiserId]
+  /// and [lineItemId].
+  ///
+  /// The list of possible [metric] values can be found here:
+  /// https://developers.google.com/bid-manager/v1.1/filters-metrics
+  ///
+  /// Throws an [ApiRequestError] if there is an error with the reporting API.
+  /// Throws an [ArgumentError] if any data is missing from the report or the
+  /// report is not found. Returns the metric value as a string.
+  Future<String> _getLineItemMetric(
+      Int64 advertiserId, Int64 lineItemId, String metric) async {
+    const dataRange = 'PREVIOUS_DAY';
+    final metrics = [metric];
+    final groupBys = [
+      'FILTER_LINE_ITEM',
+      'FILTER_DATE',
+      'FILTER_ADVERTISER_CURRENCY'
+    ];
+
+    // Creates the query for the line item.
+    final queryId = await _createLineItemQuery(
+        advertiserId, lineItemId, dataRange, metrics, groupBys);
+
+    // Get the URL of the report of the created query.
+    final url = await _getReportUrl(queryId);
+
+    try {
+      // Downloads the report and converts it into a map.
```

Comment:
Ted@ correct me if I am wrong. I talked to Jin@ previously, the reporting API won't return data directly, instead it would store your query results into a CSV file in Google Cloud Storage, so we need to download and parse the CSV file to get exact data.

Reply:
When I see "download", I think it is saving the file somewhere. But from in implementation of _reportToMap(), it seems that the csv is in the body of the GET response. So, would "retrieves" or "get" be more accurate?

Reply:
Yes, sorry that was a bad choice of wording on my part. "retrieves" makes more sense.